### PR TITLE
improve error output, stacktraces

### PR DESCRIPTION
**Pull request by eli-darkly**
I forgot that the stacktraces we were getting in error output weren't automatic - they were being added by `testify/assert` and `testify/require`. So, tests that use `matchers` instead weren't getting any stacktraces. Plus, the stacktrace output from `testify` is IMO not super attractive and includes too much of our framework code. So, I added logic to strip out their messages and add stacktrace info of our own.

The output ends up looking like this:
```
  <whatever the error message was>
  Stacktrace:
    github.com/launchdarkly/go-test-helpers/v2/matchers.AssertionScope.fail (assertion_scope.go:77)
    github.com/launchdarkly/go-test-helpers/v2/matchers.AssertionScope.Assert (assertion_scope.go:57)
    sdktests.doServerSideEventBufferTests.func4 (server_side_events_buffer.go:69)
    sdktests.doServerSideEventBufferTests (server_side_events_buffer.go:55)
    sdktests.doServerSideEventTests (testsuite_server_side.go:74)
    sdktests.RunServerSideTestSuite.func1 (testsuite_server_side.go:44)
```

The `go-test-helpers/v2/matchers` stuff should not be in there - that'll be fixed by https://github.com/launchdarkly/go-test-helpers/pull/13. Any stack frames from `github.com/launchdarkly/sdk-test-harness` will have that root package name stripped so you just get `sdktests.` etc.

Some other improvements are noted in PR comments.

### CODE REVIEW

File: framework/harness/harness.go [hunk 1]
```diff
@@ -126,11 +126,8 @@ func startServer(port int, handler http.Handler) error {
 		case <-deadline.C:
 			return fmt.Errorf("could not detect own listener at %s", server.Addr)
 		case <-ticker.C:
-			resp, err := http.DefaultClient.Head(fmt.Sprintf("http://localhost:%d", port))
-			if resp.Body != nil {
-				_ = resp.Body.Close()
-			}
-			if err == nil && resp.StatusCode == 200 {
+			_, _, err := doRequest("HEAD", fmt.Sprintf("http://localhost:%d", port), nil)
+			if err == nil {
```
Comment (eli-darkly) [inline framework/harness/harness.go:130]: This was meant to be in the previous PR where I added `doRequest`.

File: framework/harness/test_service.go [hunk 2]
```diff
@@ -148,13 +137,16 @@ func doRequest(method, url string, body []byte) ([]byte, http.Header, error) {
...
+			return nil, nil, err
+		}
 		_ = resp.Body.Close()
 	}
 	if resp.StatusCode < 200 || resp.StatusCode > 299 {
 		message := ""
-		if body != nil {
-			message = " (" + string(body) + ")"
+		if respBody != nil {
+			message = " (" + string(respBody) + ")"
```
Comment (eli-darkly) [inline framework/harness/test_service.go:149]: I was using the wrong variable here - `body` was the request body.

File: framework/harness/test_service.go [hunk 3]
```diff
@@ -203,10 +195,10 @@ func (e *TestServiceEntity) SendCommandWithParams(
 		if body == nil {
 			return errors.New("expected a response body but got none")
 		}
+		logger.Printf("Response: %s", string(body))
```
Comment (eli-darkly) [inline framework/harness/test_service.go:198]: This makes it so you can see the response body even if it failed to parse.

File: framework/ldtest/test_logger.go [hunk 4]
```diff
@@ -10,12 +10,13 @@ import (
...
-var consoleTestFailedNonCriticalColor = color.New(color.FgYellow)  //nolint:gochecknoglobals
-var consoleTestSkippedColor = color.New(color.Faint, color.FgBlue) //nolint:gochecknoglobals
-var consoleDebugOutputColor = color.New(color.Faint)               //nolint:gochecknoglobals
-var allTestsPassedColor = color.New(color.FgGreen)                 //nolint:gochecknoglobals
+var consoleTestErrorColor = color.New(color.FgYellow)                   //nolint:gochecknoglobals
+var consoleTestFailedColor = color.New(color.FgRed)                     //nolint:gochecknoglobals
+var consoleTestFailedNonCriticalColor = color.New(color.FgYellow)       //nolint:gochecknoglobals
+var consoleTestSkippedColor = color.New(color.Faint, color.FgBlue)      //nolint:gochecknoglobals
+var consoleFailedDebugOutputColor = color.New(color.Faint, color.FgRed) //nolint:gochecknoglobals
+var consolePassedDebugOutputColor = color.New(color.Faint)              //nolint:gochecknoglobals
```
Comment (eli-darkly) [inline framework/ldtest/test_logger.go:18]: Previously, the debug output for tests was always gray. Now, it'll be reddish-gray/pink for failed tests, gray for passed tests (you only see the debug output for passed tests if you specify `-debug-all`, but it's pretty verbose so now the failures will stand out more).

File: framework/ldtest/test_scope.go [hunk 5]
```diff
@@ -200,3 +206,17 @@ func (t *T) RequireCapability(name string) {
 		t.SkipWithReason(fmt.Sprintf("test service does not have capability %q", name))
 	}
 }
+
+// Helper marks the function that calls it as a test helper that shouldn't appear in stacktraces.
+// Equivalent to Go's testing.T.Helper().
```
Comment (eli-darkly) [inline framework/ldtest/test_scope.go:211]: The `testify/assert` and `testify/require` functions all call `Helper()` if it's defined, and so will the `matchers` functions once https://github.com/launchdarkly/go-test-helpers/pull/13 is merged.

File: sdktests/server_side_eval_all_flags.go [hunk 6]
```diff
@@ -180,7 +180,7 @@ func doServerSideAllFlagsExperimentationTest(t *ldtest.T) {
 		},
 		"$valid": true
 	}`
-	assert.JSONEq(t, expectedJSON, string(resultJSON))
+	m.In(t).Assert(resultJSON, m.JSONStrEqual(expectedJSON))
```
Comment (eli-darkly) [inline sdktests/server_side_eval_all_flags.go:183]: Just getting rid of the last usage of `JSONEq` because I don't like its output.

**Review by keelerm84 — APPROVED** [review]

**Pull request merged**